funsor/torch.py

```diff
@@ -139,9 +139,14 @@ def align(self, names):
         data = self.data.permute(tuple(old_dims.index(d) for d in new_dims))
         return Tensor(data, inputs, self.dtype)
 
-    def eager_subs(self, subs):
-        assert isinstance(subs, tuple)
-        subs = {k: materialize(v) for k, v in subs if k in self.inputs}
+    def eager_subs(self, dim_subs):
+        assert isinstance(dim_subs, tuple)
+        subs = {}
+        for k, v in dim_subs:
+            if k in self.inputs:
+                if not isinstance(v, Number):
+                    assert v.output != reals(), "subs for dim {} must be of bounded integer (bint) type.".format(k)
```

Comment:
Hmm, I think this would be better to do in the top level `Funsor.__call__()` to catch more errors. Let me try to put up a quick PR...

Reply:
Ok, I think this is a little safer: https://github.com/pyro-ppl/funsor/pull/62

Reply:
That looks great! Let me close this PR.